### PR TITLE
Refactored

### PR DESCRIPTION
**Pull request by sajid-ali-malik**

### CODE REVIEW

File: app/controllers/accounts_controller.rb [hunk 1]
```diff
@@ -56,4 +57,8 @@ def set_account
   def account_params
     params.require(:account).permit(:name, :account_number, :account_type, :balance)
   end
+
+  def set_transactions
+    @transactions = Transaction.where(source_account_id: @account.id).or(Transaction.where(destination_account_id: @account.id)).page(params[:page]).per(10)
```
Comment (themudassarhassan) [inline app/controllers/accounts_controller.rb]: How about 
```
@transactions = @account.transactions.page().per
```

and we will modify `Account` model in such way

```
class Account < ApplicationRecord
  ..
  ..
  def transactions
    Transaction.where(source_account_id: self.id).or(
      Transaction.where(destination_account_id: self.id)
    )
  end
end
```

File: app/controllers/budgets_controller.rb [hunk 2]
```diff
@@ -6,13 +6,12 @@ class BudgetsController < ApplicationController
 
   def index
     @budgets = policy_scope(current_user.budgets.includes(:category))
-    # TODO: Refactor this code, move budget spending calucation to budget model
-    @spent_amounts = @budgets.map { |budget| calculate_spent(budget) }
+    @spent_amounts = @budgets.map(&:spent_amount)
```
Comment (themudassarhassan) [inline app/controllers/budgets_controller.rb]: No need to create a separate instance variable, you can just call `budget.spent_amount` in views where you need to display this information.

File: app/controllers/budgets_controller.rb [hunk 3]
```diff
@@ -6,13 +6,12 @@ class BudgetsController < ApplicationController
...
   end
 
   def show
     authorize @budget
-    @spent = calculate_spent(@budget)
+    @spent = @budget.spent_amount
```
Comment (themudassarhassan) [inline app/controllers/budgets_controller.rb]: Similarly here, you can just use `budget.spent_amount`

File: app/views/home/index.html.erb [hunk 4]
```diff
@@ -47,20 +47,20 @@
...
-                <%= button_to 'Delete', budget_path(budget), method: :delete, data: { confirm: 'Are you sure?' } %>
-              </td>
-            </tr>
-          <% end %>
+            <% @budgets.each_with_index do |budget, index| %>
+              <tr>
+                <td><%= budget.category.name %></td>
+                <td><%= budget.notes %></td>
+                <td><%= budget.amount %></td>
+                <td><%= @spent_amounts[index] %></td>
```
Comment (themudassarhassan) [inline app/views/home/index.html.erb]: As mentioned [here](https://github.com/sajid-ali-malik/expense-and-budget-management-app/pull/9/files#r1680828178), you can use just `budget.spent_amount`

File: app/controllers/transactions_controller.rb [hunk 5]
```diff
@@ -2,11 +2,13 @@
 
 class TransactionsController < ApplicationController
   before_action :set_transaction, only: %i[edit update destroy show]
-  before_action :set_accounts_and_categories, only: %i[new edit update create show]
+  before_action :set_accounts_and_categories, only: %i[new edit update create show index]
 
   def index
-    @transactions = policy_scope(FetchTransactionsService.new(current_user, params).call)
-    @accounts = current_user.accounts
+    params_for_filter_service = params.permit(:source_account_id, :destination_account_id, :type,
```
Comment (themudassarhassan) [inline app/controllers/transactions_controller.rb]: Move this to a separate method.

File: app/services/fetch_transactions_service.rb [hunk 6]
```diff
@@ -1,39 +1,18 @@
...
-  def filter_by_type(transactions)
-    return transactions unless @params[:type].present?
-
-    transactions.where(type: @params[:type])
+    filter_conditions = {
+      source_account_id: @params[:source_account_id].presence,
+      destination_account_id: @params[:destination_account_id].presence,
+      category_id: @params[:category_id].presence,
+      type: @params[:type].presence
+    }.compact
```
Comment (themudassarhassan) [inline app/services/fetch_transactions_service.rb]: Move this to a separate method.

File: app/controllers/home_controller.rb [hunk 7]
```diff
@@ -6,6 +6,6 @@ def index
     @budgets = current_user.budgets.order(created_at: :desc).includes(:category).page(params[:page]).per(5)
     @transactions = current_user.transactions.order(created_at: :desc).page(params[:page]).per(5)
     @categories = Category.all
-    @spent_amounts = @budgets.map(&:spent_amount)
+    # @spent_amounts = @budgets.map(&:spent_amount)
```
Comment (themudassarhassan) [inline app/controllers/home_controller.rb:9]: Remove this.

File: app/controllers/transactions_controller.rb [hunk 8]
```diff
@@ -3,12 +3,10 @@
...
   before_action :set_transaction, only: %i[edit update destroy show]
   before_action :set_accounts_and_categories, only: %i[new edit update create show index]
+  before_action :set_params_for_filter_service, only: %i[index]
 
   def index
-    params_for_filter_service = params.permit(:source_account_id, :destination_account_id, :type,
-                                              :category_id, :sort, :page)
-
-    @transactions = policy_scope(FetchTransactionsService.new(current_user, params_for_filter_service).call)
+    @transactions = policy_scope(FetchTransactionsService.new(current_user, params).call)
```
Comment (themudassarhassan) [inline app/controllers/transactions_controller.rb:9]: Is it working?

Comment (sajid-ali-malik) [reply app/controllers/transactions_controller.rb:9]: yes

**Pull request merged**